### PR TITLE
[DTensor][BE] redistribute to replicate in from_local backward for partial target type

### PR DESCRIPTION
**Pull request by acisseJZhong**
follow up per this [comment](https://github.com/pytorch/pytorch/pull/170147#discussion_r2616173713) 

**Motivation:** If in `from_local` forward placement is Partial, user could get either Partial or Replicate gradients depending on what the incoming gradient placement is(if incoming gradient is Replicate, we keep Replicate; if incoming gradient is Partial, we keep Partial). This is not ideal as it's not 1-1 mapping and introduces ambiguity. 

**Change in this PR:** if incoming gradient is Partial, we redistribute the Partial to Replicate. As a result, users will always get Replicate gradient when forward is Partial. 

This behavior aligned with our future plan to add Reduced/Unreduced type: If forward is partial, backward is Replicate(now)/Reduced(later). 

Also added unit test for `from_local` backward to cover the 5 possible cases:
| `from_local` forward | grad_output | grad_input |
|-----|-----|-----|
| Shard | Shard | Shard |
| Replicate | Replicate | Replicate | 
| Replicate | Partial | Replicate | 
| Partial | Replicate | Replicate(now) / Reduced(later) | 
| Partial | Partial | Replicate(now) / Reduced(later) | 
```
python3 test/distributed/tensor/test_dtensor.py -k test_from_local_backward
```

### CODE REVIEW

File: torch/distributed/tensor/_api.py [hunk 1]
```diff
@@ -210,20 +209,16 @@ def backward(ctx, grad_output: "DTensor"):  # type: ignore[override]
         if grad_output.placements != forward_input_placements:
             current_spec = grad_output._spec
 
-            # for backward shard -> partial, we do shard -> replicate
-            # for backward replicate -> partial, we do replicate -> replicate / skip transformation
-            # TODO: for backward partial -> partial, right now we keep it unchanged,
-            # but this might need revisit
+            # if target gradient placement is partial, we always redistribute to replicate, this means
```
Comment (vishal9-team) [inline torch/distributed/tensor/_api.py]: 1. Can the comment clarify this in-terms of gradient mapping guarantee that this function now provides in the backward? (Shard-Shard, Replicate-Replicate, Partial-Replicate) - maybe even as a table for clarity

2. It'll be useful to highlight that this particular example will be BC breaking for the case of 

OLD: Partial(fwd) - Partial (grad_out) - Partial (grad_in)
NEW: Partial(fwd) - Partial (grad_out) - Replicate (grad_in),

File: torch/distributed/tensor/_api.py [hunk 2]
```diff
@@ -416,18 +411,23 @@ def from_local(
...
-                placements must use the same reduce operation.
-
         .. note:: When ``run_check=False``, it is the user's responsibility to ensure the
             local tensor passed in is correct across ranks (i.e. the tensor is sharded for
             the ``Shard(dim)`` placement or replicated for the ``Replicate()`` placement).
             If not, the behavior of the created DTensor is undefined.
 
         .. note:: ``from_local`` is differentiable, the `requires_grad` of the created
             `DTensor` object will depend on if `local_tensor` requires_grad or not.
+
```
Comment (vishal9-team) [inline torch/distributed/tensor/_api.py:423]: Is it possible to have some tests for this?

File: torch/distributed/tensor/_api.py (hunk 1 above)
Comment (acisseJZhong) [reply torch/distributed/tensor/_api.py]: good idea, added a table to clarity in the doc string

File: torch/distributed/tensor/_api.py (hunk 2 above)
Comment (acisseJZhong) [reply torch/distributed/tensor/_api.py:423]: sure, let me add tests for each case! although according to yesterday's discussion we are unlikely to hit `Partial(fwd) - Partial(grad_output)` case?

File: test/distributed/tensor/test_dtensor.py [hunk 3]
```diff
@@ -222,6 +222,74 @@ def test_from_local(self):
...
+            output = dt * 2
+            grad = DTensor.from_local(
+                torch.ones(3, 3, device=self.device_type), device_mesh, [Shard(0)]
+            )
+            output.backward(grad)
+            self.assertEqual(local_tensor.grad, torch.ones(3, 3) * 2)
+        # Gradient should flow back without redistribution
+        self.assertEqual(comm_mode.get_total_counts(), 0)
+
+        # Test 2: Replicate -> Replicate (no redistribution needed in backward)
```
Comment (tianyu-l) [inline test/distributed/tensor/test_dtensor.py]: add a test case where forward is Replicate, backward is Partial -> Replicate

File: test/distributed/tensor/test_dtensor.py [hunk 4]
```diff
@@ -222,6 +222,74 @@ def test_from_local(self):
...
+                device_mesh,
+                [Replicate()],
+                run_check=False,
+            )
+            output.backward(grad)
+            self.assertEqual(local_tensor.grad, torch.ones(3, 3) * 2)
+        # Gradient should flow back without redistribution
+        self.assertEqual(comm_mode.get_total_counts(), 0)
+
+        # Test 3: Partial -> Replicate (gradient normalized to Replicate)
```
Comment (tianyu-l) [inline test/distributed/tensor/test_dtensor.py]: add a test case where forward is Partial, backward is Replicate -> keep Replicate

**Review by tianyu-l — APPROVED** [review]
sgtm, please address remaining comments.

File: test/distributed/tensor/test_dtensor.py (hunk 4 above)
Comment (acisseJZhong) [reply test/distributed/tensor/test_dtensor.py]: oh right! now I finally understand the full picture. The 5 test cases should cover all the possibilities.

File: torch/distributed/tensor/_api.py (hunk 2 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_api.py:423]: We should test that grad_output correctly gets redistributed to Replicate in this case
